Fix early output notification

eed3si9n opened the pull request:
Fixes https://github.com/sbt/zinc/issues/847

This fixes the no-op case.

<s>
This fixes 2 related bugs.

First is that prior to this change, early notification could have triggered multiple times in `AnalysisCallback#getCycleResultOnce`. This fixes that by moving that to `CycleState#completeCycles`, which is called only for the final cycle.
</s>

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/849/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

dwijnand commented on internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala:
```diff
@@ -69,12 +75,32 @@ private[inc] abstract class IncrementalCommon(
-  ) {
+      cycleNum: Int,
+      progress: Option[CompileProgress],
+      var notifiedEarly: Boolean = false,
```
Ah, damn, I don't like this detail.  It's already very confusingly callback hell and recursive that I really don't want to add mutation to the mix...

I would say let's drop the (never-was-documented) guarantee that `earlyOutputComplete` is only called once, and let the implementation deal with it.  `tryComplete` on `Promise` doesn't break like `complete`.

eed3si9n replied:
It's not only about once or multiple times that's the problem but also when. Prior to this change the notification was happening either prematurely before the final cycle or sometimes not triggered at all. We can refractor the impl details but this fix needs to happen.

eed3si9n replied:
Reading the code again, maybe I misunderstood, and I might need to only handle the no-op cases. If so I can get rid of the var.

eed3si9n replied:
https://github.com/sbt/zinc/pull/849/commits/75a3bb91785fed6326de4b48a0d8653eb994fecf is a much smaller change I just pushed.

dwijnand commented on zinc/src/main/scala/sbt/internal/inc/IncrementalCompilerImpl.scala:
```diff
@@ -469,7 +469,7 @@ class IncrementalCompilerImpl extends IncrementalCompiler {
       )
       def prevResult = CompileResult.of(prev, config.currentSetup, false)
-      if (skip) prevResult
+      if (skip && !incrementalOptions.pipelining) prevResult
```
Why this change?

eed3si9n replied:
This is so the notification would fire up.

eed3si9n committed:
Fix early output notification

Fixes #847

This fixes the early output notification in the no-op case.

eatkins commented on internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala:
```diff
@@ -382,10 +382,10 @@ private[inc] abstract class IncrementalCommon(
-        val changed0 = new java.util.HashSet[VirtualFileRef]
-        val removed0 = new java.util.HashSet[VirtualFileRef]
-        val unmodified0 = new java.util.HashSet[VirtualFileRef]
+        val added0 = java.util.concurrent.ConcurrentHashMap.newKeySet[VirtualFileRef]
```
Thanks for addressing this! I'm not going to review the PR because I have no context on the other commit.

eed3si9n replied:
I'm going to split that into a separate PR - https://github.com/sbt/zinc/pull/850

eed3si9n commented:
source-dependencies/subproject-pipelining still failing on Windows - https://ci.appveyor.com/project/sbt/zinc/builds/34246755

<details>
<summary>

```scala
[error] ${BASE}/use/B.scala:4:11: not found: value A
[error]   val y = A.x
[error]           ^
[error] one error found
....
[error] x source-dependencies/subproject-pipelining
[error] sbt.internal.inc.BatchScriptRunner$PreciseScriptedError: {line 2} Command failed: 'use/compile'
```

</summary>

```scala
[info] Running source-dependencies/subproject-pipelining
[debug] Reading incremental options from map
[debug] PIPELINING value was read.
[debug] Reading incremental options from map
[debug] PIPELINING value was read.
[debug] [zinc] IncrementalCompile -----------
[debug] IncrementalCompile.incrementalCompile
[debug] previous = Stamps for: 0 products, 0 sources, 0 libraries Map()
[debug] current source = Set(${BASE}/dep/A.scala)
[debug] > initialChanges = InitialChanges(Changes(added = Set(${BASE}/dep/A.scala), removed = Set(), changed = Set(), unmodified = ...),Set(),Set(),API Changes: Set())
[debug] Full compilation, no sources in previous analysis.
[debug] all 1 sources are invalidated
[debug] Initial set of included nodes: 
[debug] compilation cycle 1
[info] Compiling 1 Scala source to C:\Users\appveyor\AppData\Local\Temp\1\sbt_18d5a6eb\dep\target\classes ...
[debug] Returning already retrieved and compiled bridge: C:\Users\appveyor\AppData\Local\Temp\1\sbt_3ff58411\target-bridge-2.13.3.jar.
[debug] [zinc] Running cached compiler 3e575a50 for Scala compiler version 2.13.3
[debug] [zinc] The Scala compiler is invoked with:
[debug] 	-Ypickle-java
[debug] 	-Ypickle-write
[debug] 	C:\Users\appveyor\AppData\Local\Temp\1\sbt_18d5a6eb\dep\target\early\output.jar
[debug] 	-bootclasspath
[debug] 	C:\Program Files\Java\jdk1.8.0\jre\lib\resources.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\rt.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\sunrsasign.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\jsse.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\jce.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\charsets.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\jfr.jar;C:\Program Files\Java\jdk1.8.0\jre\classes;C:\Users\appveyor\AppData\Local\Coursier\cache\v1\https\repo1.maven.org\maven2\org\scala-lang\scala-library\2.13.3\scala-library-2.13.3.jar
[debug] 	-classpath
[debug] 	C:\Users\appveyor\AppData\Local\Coursier\cache\v1\https\repo1.maven.org\maven2\org\scala-lang\scala-compiler\2.13.3\scala-compiler-2.13.3.jar;C:\Users\appveyor\AppData\Local\Coursier\cache\v1\https\repo1.maven.org\maven2\org\scala-lang\scala-reflect\2.13.3\scala-reflect-2.13.3.jar;C:\Users\appveyor\AppData\Local\Coursier\cache\v1\https\repo1.maven.org\maven2\org\jline\jline\3.15.0\jline-3.15.0.jar;C:\Users\appveyor\AppData\Local\Coursier\cache\v1\https\repo1.maven.org\maven2\net\java\dev\jna\jna\5.3.1\jna-5.3.1.jar;C:\Users\appveyor\AppData\Local\Temp\1\sbt_18d5a6eb\dep\target\classes
[info] [progress] early output is done for dep!
[debug] Scala compilation took 0.4920256 s
[debug] done compiling Scala sources
[info] dep: compilation done: C:\Users\appveyor\AppData\Local\Temp\1\sbt_18d5a6eb\dep\A.scala
[info] dep: compilation done: C:\Users\appveyor\AppData\Local\Temp\1\sbt_18d5a6eb\dep\A.scala
[debug] [zinc] IncrementalCompile -----------
[debug] IncrementalCompile.incrementalCompile
[debug] previous = Stamps for: 0 products, 0 sources, 0 libraries Map()
[debug] current source = Set(${BASE}/use/B.scala)
[debug] > initialChanges = InitialChanges(Changes(added = Set(${BASE}/use/B.scala), removed = Set(), changed = Set(), unmodified = ...),Set(),Set(),API Changes: Set())
[debug] Full compilation, no sources in previous analysis.
[debug] all 1 sources are invalidated
[debug] Initial set of included nodes: 
[debug] compilation cycle 1
[info] Compiling 1 Scala source to C:\Users\appveyor\AppData\Local\Temp\1\sbt_18d5a6eb\use\target\classes ...
[debug] Returning already retrieved and compiled bridge: C:\Users\appveyor\AppData\Local\Temp\1\sbt_3ff58411\target-bridge-2.13.3.jar.
[debug] [zinc] Running cached compiler 30305177 for Scala compiler version 2.13.3
[debug] [zinc] The Scala compiler is invoked with:
[debug] 	-Ypickle-java
[debug] 	-Ypickle-write
[debug] 	C:\Users\appveyor\AppData\Local\Temp\1\sbt_18d5a6eb\use\target\early\output.jar
[debug] 	-bootclasspath
[debug] 	C:\Program Files\Java\jdk1.8.0\jre\lib\resources.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\rt.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\sunrsasign.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\jsse.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\jce.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\charsets.jar;C:\Program Files\Java\jdk1.8.0\jre\lib\jfr.jar;C:\Program Files\Java\jdk1.8.0\jre\classes;C:\Users\appveyor\AppData\Local\Coursier\cache\v1\https\repo1.maven.org\maven2\org\scala-lang\scala-library\2.13.3\scala-library-2.13.3.jar
[debug] 	-classpath
[debug] 	C:\Users\appveyor\AppData\Local\Coursier\cache\v1\https\repo1.maven.org\maven2\org\scala-lang\scala-compiler\2.13.3\scala-compiler-2.13.3.jar;C:\Users\appveyor\AppData\Local\Coursier\cache\v1\https\repo1.maven.org\maven2\org\scala-lang\scala-reflect\2.13.3\scala-reflect-2.13.3.jar;C:\Users\appveyor\AppData\Local\Coursier\cache\v1\https\repo1.maven.org\maven2\org\jline\jline\3.15.0\jline-3.15.0.jar;C:\Users\appveyor\AppData\Local\Coursier\cache\v1\https\repo1.maven.org\maven2\net\java\dev\jna\jna\5.3.1\jna-5.3.1.jar;C:\Users\appveyor\AppData\Local\Temp\1\sbt_18d5a6eb\use\target\classes;C:\Users\appveyor\AppData\Local\Temp\1\sbt_18d5a6eb\dep\target\early\output.jar
[error] ${BASE}/use/B.scala:4:11: not found: value A
[error]   val y = A.x
[error]           ^
[error] one error found
[debug] Compilation failed (CompilerInterface)
[error] ## Exception when compiling 1 sources to C:\Users\appveyor\AppData\Local\Temp\1\sbt_18d5a6eb\use\target\classes
[error] Compilation failed
[error] xsbt.CachedCompiler0.handleErrors(CompilerInterface.scala:179)
[error] xsbt.CachedCompiler0.run(CompilerInterface.scala:168)
[error] xsbt.CachedCompiler0.run(CompilerInterface.scala:130)
[error] xsbt.CompilerInterface.run(CompilerInterface.scala:38)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:86)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:176)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:231)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$5(MixedAnalyzingCompiler.scala:167)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$5$adapted(MixedAnalyzingCompiler.scala:156)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:230)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:156)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:184)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:570)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:570)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$4(Incremental.scala:166)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$4$adapted(Incremental.scala:164)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:410)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:121)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:288)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$9(Incremental.scala:379)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:450)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:366)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:158)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:570)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:418)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.internal.inc.ProjectStructure.doCompile(IncHandler.scala:641)
[error] sbt.internal.inc.ProjectStructure.$anonfun$startCompilation$6(IncHandler.scala:539)
[error] scala.util.Success.$anonfun$map$1(Try.scala:255)
[error] scala.util.Success.map(Try.scala:213)
[error] scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
[error] scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
[error] scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
[error] scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
[error] java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
[error] java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
[error] java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
[error] java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
[error] java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
[error]            
[error] x source-dependencies/subproject-pipelining
[error] sbt.internal.inc.BatchScriptRunner$PreciseScriptedError: {line 2} Command failed: 'use/compile'
[error] Caused by: Compilation failed
[error] 	at xsbt.CachedCompiler0.handleErrors(CompilerInterface.scala:179)
[error] 	at xsbt.CachedCompiler0.run(CompilerInterface.scala:168)
[error] 	at xsbt.CachedCompiler0.run(CompilerInterface.scala:130)
[error] 	at xsbt.CompilerInterface.run(CompilerInterface.scala:38)
[error] 	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:86)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:176)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:231)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$5(MixedAnalyzingCompiler.scala:167)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$5$adapted(MixedAnalyzingCompiler.scala:156)
[error] 	at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:230)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:156)
[error] 	at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:184)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:570)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:570)
[error] 	at sbt.internal.inc.Incremental$.$anonfun$apply$4(Incremental.scala:166)
[error] 	at sbt.internal.inc.Incremental$.$anonfun$apply$4$adapted(Incremental.scala:164)
[error] 	at sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:410)
[error] 	at sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:121)
[error] 	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] 	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] 	at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:288)
[error] 	at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$9(Incremental.scala:379)
[error] 	at sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:450)
[error] 	at sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:366)
[error] 	at sbt.internal.inc.Incremental$.apply(Incremental.scala:158)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:570)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:418)
[error] 	at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] 	at sbt.internal.inc.ProjectStructure.doCompile(IncHandler.scala:641)
[error] 	at sbt.internal.inc.ProjectStructure.$anonfun$startCompilation$6(IncHandler.scala:539)
[error] 	at scala.util.Success.$anonfun$map$1(Try.scala:255)
[error] 	at scala.util.Success.map(Try.scala:213)
[error] 	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
[error] 	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
[error] 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
[error] 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
[error] 	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
[error] 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
[error] 	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
[error] 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
[error] 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

</details>

eed3si9n committed:
Assert in scripted that the early output exists

eed3si9n commented:
Testing this on a Windows laptop, and it seems like the early JAR is missing sig files.
I think this is a regression in https://github.com/sbt/zinc/pull/821 since `Set[String]` is `Set("example\A.class", "example\A$.class")` (note the backslash), and it's deleting `/example/A.sig`.

eed3si9n committed:
Reproduce 851

https://github.com/sbt/zinc/issues/851

eed3si9n committed:
Convert known products into slash path

Fixes https://github.com/sbt/zinc/issues/851

This normalizes backslash to slash / in parallel.

eed3si9n commented:
https://github.com/sbt/zinc/pull/852 is green so I'm going to rebase this on that.

eed3si9n committed:
Rename to CompileProgress#afterEarlyOutput
